Feature #26 카테고리 등록 상품 개수 읽기

KJW921222 opened the pull request:
1. 카테고리 별 등록 상품개수 조회 기능을 구현한다.

KJW921222 committed:
commit #1

KJW921222 committed:
## 개요
1. 카테고리 별 등록 상품개수 조회 기능 작성
2. RequestDTO에 통일성을 주기 위해 불필요한 메소드 삭제

## 작업사항
1. 카테고리 별 등록 상품개수 조회 기능 및 테스트 코드 작성
2. ItemDetailRequestDTO의 불필요한 메소드 삭제, ItemService 수정

KJW921222 committed:
## 개요
1. 카테고리 별 등록 상품개수 조회 기능 작성
2. RequestDTO에 통일성을 주기 위해 불필요한 메소드 삭제

## 작업사항
1. 카테고리 별 등록 상품개수 조회 기능 및 테스트 코드 작성
2. ItemDetailRequestDTO의 불필요한 메소드 삭제, ItemService 수정

KJW921222 commented:
@f-lab-robert

f-lab-robert commented on src/test/java/com/flabedu/small/small/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,53 @@
+                        .categoryId(2l).parentId(null).name("하의").itemCount(2).build()
+        );
+
+        List<CategoryResponseDTO> test = List.of(
```
이 부분은 전혀 사용되고 있지 않는것 같은데요.

f-lab-robert commented on src/test/java/com/flabedu/small/small/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,53 @@
+    CategoryMapper categoryMapper;
+
+    @Test
+    @DisplayName("카테고리 별 상품 개수 조회 시 예상 값이 반환된다")
```
예상하지 않은 값이 반환 되는 가령 DAO 에서 예상치 못한 에러가 발생했을 경우에 대한 테스트 케이스도 필요하지 않을까요 ? 
현재는 nice case 에 대한 상황만 고려가 되는 것 같습니다.

KJW921222 replied:
@f-lab-robert 
sql의 결과가 없다거나, db와의 연결이 끊어졌을 때라고 생각해도 될까요?

f-lab-robert replied:
@KJW921222 네 그러한 예외 상황들에 대한 고민 및 테스트 케이스가 필요해보입니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,65 @@
+    CategoryResponseDTO testDto1;
+
+    @BeforeEach
+    void setup() {
```
현재 테스트 하려는 함수가 1개인것 같은데요. 이 부분은 제거하고 실제 데이터 셋팅하는 부분을 findCategorySuccess 로 옮겨보시겠어요 ? 

추후에 작성하는 테스트 코드에 다음과 관련한 로직이 반복 되었을 경우에 다음과 같이 데이터 셋업하는 걸 고려해봐도 좋을 것 같습니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Test
+    @DisplayName("카테고리 별 상품 개수 조회 시 예상한 값이 반환된다")
```
예상한 값이 반환되지 않는 경우도 있을까요 ?

f-lab-robert commented on src/main/resources/mapper/CategoryMapper.xml:
```diff
@@ -0,0 +1,37 @@
+
+    <select id="findCategory" resultType="com.flabedu.small.small.dao.CategoryItemCountDao">
+        SELECT a.category_id,a.parent as parentId,a.name
+            ,(case when a.parent is null then ifnull(parentcount,0) else ifnull(itemcount,0) end) itemCount
```
현재는 sql 에 굉장히 종속적인 코드가 되어가고 있는 것 같은데요. 
차후에 작업하실 때는 sql 을 2번 날리고 business logic 에서 필요 정보를 조립하는 방식으로 하는건 어떨지 그에 대한 장단점은 무엇이 있을지도 생각해봐주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/controller/CategoryController.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequestMapping
```
이 부분은 어떤 의미로 현재 사용하고 계신가요 ? 아무 역할이 없어보이는데요.

KJW921222 committed:
## 개요
1. CategoryServiceTest 테스트 케이스 추가

## 작업사항
1. CategoryServiceTest 테스트 케이스 추가
  - 등록된 카테고리가 없는 경우
  - DB에서 에러가 발생한 경우

KJW921222 committed:
## 개요
1. CategoryServiceTest 테스트 케이스 추가

## 작업사항
1. CategoryServiceTest 테스트 케이스 추가
  - 등록된 카테고리가 없는 경우
  - DB에서 에러가 발생한 경우
